File: plover_vm/memory/apu.py
"""APU PSG state for Mailbox audio commands (4ch square/noise)."""

from __future__ import annotations

from dataclasses import dataclass, field

CMD_APU_SET_CTRL = 0x50
CMD_APU_CH_WRITE = 0x51
CMD_APU_CH_SYNC = 0x52
CMD_APU_CH_OFF = 0x53

APU_CHANNELS = 4
NOISE_CHANNEL = 3

WAVE_OFF = 0
WAVE_SQUARE = 1
WAVE_NOISE = 2

CLK_HZ = 44_100
SAMPLE_RATE = 22_050


@dataclass
class Channel:
    period: int = 0
    volume: int = 0
    wave: int = WAVE_OFF

    def copy(self) -> Channel:
        return Channel(period=self.period, volume=self.volume, wave=self.wave)


@dataclass
class ApuState:
    """4ch PSG; staged writes commit on APU_CH_SYNC."""

    channels: list[Channel] = field(default_factory=lambda: [Channel() for _ in range(APU_CHANNELS)])
    pending: list[Channel] = field(default_factory=lambda: [Channel() for _ in range(APU_CHANNELS)])
    master_vol: int = 15
    mute: bool = False
    beep_ticks: int = 0
    clk_hz: int = CLK_HZ
    sample_rate: int = SAMPLE_RATE
    _phase: list[int] = field(default_factory=lambda: [0] * APU_CHANNELS)
    _lfsr: int = 0x7FFF

    def set_ctrl(self, buffer: bytearray | bytes) -> bool:
        if len(buffer) < 2:
            return False
        self.master_vol = buffer[0] & 0x0F
        self.mute = bool(buffer[1] & 0x01)
        return True

    def stage_ch_write(self, buffer: bytearray | bytes) -> bool:
        if len(buffer) < 5:
            return False
        ch = buffer[0] & 0x03
        period = buffer[1] | (buffer[2] << 8)
        vol = buffer[3] & 0x0F
        wave = buffer[4] & 0x03
        if ch == NOISE_CHANNEL:
            if wave not in (WAVE_OFF, WAVE_NOISE):
                return False
        elif wave not in (WAVE_OFF, WAVE_SQUARE):
            return False
        self.pending[ch] = Channel(period=period & 0xFFFF, volume=vol, wave=wave)
        return True

    def sync(self) -> bool:
        for i in range(APU_CHANNELS):
            self.channels[i] = self.pending[i].copy()
            self._phase[i] = 0
        return True

    def ch_off(self, ch: int) -> bool:
        if ch < 0 or ch >= APU_CHANNELS:
            return False
        off = Channel()
        self.channels[ch] = off.copy()
        self.pending[ch] = off.copy()
        self._phase[ch] = 0
        return True

    def dispatch(self, cmd: int, param: int, buffer: bytearray) -> bool:
        """Return True if command accepted, False if silent drop."""
        if cmd == CMD_APU_SET_CTRL:
            return self.set_ctrl(buffer)
        if cmd == CMD_APU_CH_WRITE:
            return self.stage_ch_write(buffer)
        if cmd == CMD_APU_CH_SYNC:
            return self.sync()
        if cmd == CMD_APU_CH_OFF:
            return self.ch_off(param & 0x03)
        return False

    def freq_hz(self, ch: int) -> float:
        c = self.channels[ch]
        if c.period <= 0 or c.wave == WAVE_OFF:
            return 0.0
        return self.clk_hz / (2.0 * c.period)

    def mix_samples(self, n: int) -> bytes:
        """Generate n mono 8-bit samples (center 128) for VM tests."""
        out = bytearray(n)
        if self.mute:
            return bytes([128] * n)
        masters = max(1, self.master_vol) / 15.0
        tick_step = self.clk_hz // self.sample_rate
        for i in range(n):
            acc = 0.0
            for ch in range(APU_CHANNELS):
                c = self.channels[ch]
                if c.wave == WAVE_OFF or c.volume == 0 or c.period <= 0:
                    continue
                if c.wave == WAVE_SQUARE:
                    tick = self._phase[ch]
                    if ((tick // c.period) % 2) == 0:
                        acc += c.volume
                    self._phase[ch] = (tick + tick_step) % (c.period * 2)
                elif c.wave == WAVE_NOISE and ch == NOISE_CHANNEL:
                    bit = self._lfsr & 1
                    self._lfsr = (self._lfsr >> 1) | ((bit ^ ((self._lfsr >> 1) & 1)) << 14)
                    if bit:
                        acc += c.volume
            level = 128 + int((acc / (APU_CHANNELS * 15.0)) * 127.0 * masters)
            out[i] = max(0, min(255, level)) & 0xFF
        return bytes(out)

    def zero_crossings(self, samples: bytes) -> int:
        """Count approximate zero-crossings around 128 for tone tests."""
        crosses = 0
        prev = samples[0] - 128
        for b in samples[1:]:
            cur = b - 128
            if (prev > 0 and cur <= 0) or (prev <= 0 and cur > 0):
                crosses += 1
            prev = cur
        return crosses

File: crates/plover_copro/src/apu.rs
pub const CMD_APU_SET_CTRL: u8 = 0x50;
pub const CMD_APU_CH_WRITE: u8 = 0x51;
pub const CMD_APU_CH_SYNC: u8 = 0x52;
pub const CMD_APU_CH_OFF: u8 = 0x53;

pub const APU_CHANNELS: usize = 4;
pub const NOISE_CHANNEL: usize = 3;

pub const WAVE_OFF: u8 = 0;
pub const WAVE_SQUARE: u8 = 1;
pub const WAVE_NOISE: u8 = 2;

pub const CLK_HZ: u32 = 44_100;
pub const SAMPLE_RATE: u32 = 22_050;

#[derive(Clone, Copy, Default, Debug, PartialEq, Eq)]
pub struct Channel {
    pub period: u16,
    pub volume: u8,
    pub wave: u8,
}

#[derive(Clone, Debug)]
pub struct ApuState {
    pub channels: [Channel; APU_CHANNELS],
    pub pending: [Channel; APU_CHANNELS],
    pub master_vol: u8,
    pub mute: bool,
    pub beep_ticks: u16,
    pub clk_hz: u32,
    pub sample_rate: u32,
    phase: [u32; APU_CHANNELS],
    lfsr: u16,
}

impl Default for ApuState {
    fn default() -> Self {
        Self {
            channels: [Channel::default(); APU_CHANNELS],
            pending: [Channel::default(); APU_CHANNELS],
            master_vol: 15,
            mute: false,
            beep_ticks: 0,
            clk_hz: CLK_HZ,
            sample_rate: SAMPLE_RATE,
            phase: [0; APU_CHANNELS],
            lfsr: 0x7FFF,
        }
    }
}

impl ApuState {
    pub fn set_ctrl(&mut self, buffer: &[u8]) -> bool {
        if buffer.len() < 2 {
            return false;
        }
        self.master_vol = buffer[0] & 0x0F;
        self.mute = (buffer[1] & 0x01) != 0;
        true
    }

    pub fn stage_ch_write(&mut self, buffer: &[u8]) -> bool {
        if buffer.len() < 5 {
            return false;
        }
        let ch = (buffer[0] & 0x03) as usize;
        let period = u16::from(buffer[1]) | (u16::from(buffer[2]) << 8);
        let vol = buffer[3] & 0x0F;
        let wave = buffer[4] & 0x03;
        if ch == NOISE_CHANNEL {
            if wave != WAVE_OFF && wave != WAVE_NOISE {
                return false;
            }
        } else if wave != WAVE_OFF && wave != WAVE_SQUARE {
            return false;
        }
        self.pending[ch] = Channel {
            period,
            volume: vol,
            wave,
        };
        true
    }

    pub fn sync(&mut self) -> bool {
        for i in 0..APU_CHANNELS {
            self.channels[i] = self.pending[i];
            self.phase[i] = 0;
        }
        true
    }

    pub fn ch_off(&mut self, ch: usize) -> bool {
        if ch >= APU_CHANNELS {
            return false;
        }
        let off = Channel::default();
        self.channels[ch] = off;
        self.pending[ch] = off;
        self.phase[ch] = 0;
        true
    }

    pub fn dispatch(&mut self, cmd: u8, param: u8, buffer: &mut [u8; 248]) -> bool {
        match cmd {
            CMD_APU_SET_CTRL => self.set_ctrl(buffer),
            CMD_APU_CH_WRITE => self.stage_ch_write(buffer),
            CMD_APU_CH_SYNC => self.sync(),
            CMD_APU_CH_OFF => self.ch_off((param & 0x03) as usize),
            _ => false,
        }
    }

    pub fn mix_samples(&mut self, n: usize) -> Vec<u8> {
        let mut out = vec![128u8; n];
        if self.mute {
            return out;
        }
        let masters = (self.master_vol.max(1) as f32) / 15.0;
        let tick_step = self.clk_hz / self.sample_rate;
        for sample in out.iter_mut() {
            let mut acc = 0.0f32;
            for ch in 0..APU_CHANNELS {
                let c = &self.channels[ch];
                if c.wave == WAVE_OFF || c.volume == 0 || c.period == 0 {
                    continue;
                }
                if c.wave == WAVE_SQUARE {
                    let tick = self.phase[ch];
                    if c.period > 0 && ((tick / u32::from(c.period)) % 2) == 0 {
                        acc += f32::from(c.volume);
                    }
                    let period2 = u32::from(c.period) * 2;
                    self.phase[ch] = if period2 > 0 {
                        (tick + tick_step) % period2
                    } else {
                        0
                    };
                } else if c.wave == WAVE_NOISE && ch == NOISE_CHANNEL {
                    let bit = self.lfsr & 1;
                    self.lfsr = (self.lfsr >> 1)
                        | (((bit ^ ((self.lfsr >> 1) & 1)) as u16) << 14);
                    if bit != 0 {
                        acc += f32::from(c.volume);
                    }
                }
            }
            let level = 128.0 + (acc / (APU_CHANNELS as f32 * 15.0)) * 127.0 * masters;
            *sample = level.round().clamp(0.0, 255.0) as u8;
        }
        out
    }

    pub fn zero_crossings(samples: &[u8]) -> usize {
        let mut crosses = 0;
        let mut prev = i16::from(samples[0]) - 128;
        for &b in &samples[1..] {
            let cur = i16::from(b) - 128;
            if (prev > 0 && cur <= 0) || (prev <= 0 && cur > 0) {
                crosses += 1;
            }
            prev = cur;
        }
        crosses
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn buf5(data: [u8; 5]) -> [u8; 248] {
        let mut b = [0u8; 248];
        b[..5].copy_from_slice(&data);
        b
    }

    #[test]
    fn apu_set_ctrl() {
        let mut apu = ApuState::default();
        assert!(apu.dispatch(CMD_APU_SET_CTRL, 0, &mut buf5([10, 0, 0, 0, 0])));
        assert_eq!(apu.master_vol, 10);
        assert!(!apu.mute);
        assert!(apu.dispatch(CMD_APU_SET_CTRL, 0, &mut buf5([5, 1, 0, 0, 0])));
        assert_eq!(apu.master_vol, 5);
        assert!(apu.mute);
    }

    #[test]
    fn apu_ch_write_sync() {
        let mut apu = ApuState::default();
        assert!(apu.dispatch(CMD_APU_CH_WRITE, 0, &mut buf5([0, 22, 0, 15, WAVE_SQUARE])));
        assert_eq!(apu.channels[0].period, 0);
        assert!(apu.dispatch(CMD_APU_CH_SYNC, 0, &mut [0u8; 248]));
        assert_eq!(apu.channels[0].period, 22);
        assert_eq!(apu.channels[0].volume, 15);
    }

    #[test]
    fn apu_ch_off() {
        let mut apu = ApuState::default();
        apu.dispatch(CMD_APU_CH_WRITE, 0, &mut buf5([0, 30, 0, 10, WAVE_SQUARE]));
        apu.dispatch(CMD_APU_CH_SYNC, 0, &mut [0u8; 248]);
        assert!(apu.dispatch(CMD_APU_CH_OFF, 0, &mut [0u8; 248]));
        assert_eq!(apu.channels[0].period, 0);
        assert_eq!(apu.channels[0].wave, WAVE_OFF);
    }

    #[test]
    fn apu_ch3_noise_only() {
        let mut apu = ApuState::default();
        assert!(apu.dispatch(CMD_APU_CH_WRITE, 0, &mut buf5([3, 10, 0, 8, WAVE_NOISE])));
        assert!(!apu.dispatch(CMD_APU_CH_WRITE, 0, &mut buf5([3, 10, 0, 8, WAVE_SQUARE])));
        apu.dispatch(CMD_APU_CH_SYNC, 0, &mut [0u8; 248]);
        assert_eq!(apu.channels[3].wave, WAVE_NOISE);
    }
}
